ingest/src/main/java/com/microsoft/azure/kusto/ingest/IngestClientBase.java

```diff
@@ -28,4 +36,128 @@
         }
     }
 
+    protected abstract IngestionResult ingestFromFileImpl(FileSourceInfo fileSourceInfo, IngestionProperties ingestionProperties)
+            throws IngestionClientException, IngestionServiceException;
+
+    public IngestionResult ingestFromFile(FileSourceInfo fileSourceInfo, IngestionProperties ingestionProperties)
+            throws IngestionClientException, IngestionServiceException {
+        // trace ingestFromFile
+        try (DistributedTracing.Span span = DistributedTracing.startSpan(getClientType().concat(".ingestFromFile"), Context.NONE, ProcessKind.PROCESS,
+                getIngestionTraceAttributes(fileSourceInfo, ingestionProperties))) {
+            try {
+                return ingestFromFileImpl(fileSourceInfo, ingestionProperties);
+            } catch (IngestionClientException | IngestionServiceException e) {
+                span.addException(e);
+                throw e;
+            }
+        }
+    }
+
+    /**
+     * <p>Ingest data from a blob storage into Kusto database.</p>
+     * This method ingests the data from a given blob, described in {@code blobSourceInfo}, into Kusto database,
+     * according to the properties mentioned in {@code ingestionProperties}
+     *
+     * @param blobSourceInfo      The specific SourceInfo to be ingested
+     * @param ingestionProperties Settings used to customize the ingestion operation
+     * @return {@link IngestionResult} object including the ingestion result
+     * @throws IngestionClientException  An exception originating from a client activity
+     * @throws IngestionServiceException An exception returned from the service
+     * @see BlobSourceInfo
+     * @see IngestionProperties
+     */
+    protected abstract IngestionResult ingestFromBlobImpl(BlobSourceInfo blobSourceInfo, IngestionProperties ingestionProperties)
+            throws IngestionClientException, IngestionServiceException;
+
+    public IngestionResult ingestFromBlob(BlobSourceInfo blobSourceInfo, IngestionProperties ingestionProperties)
+            throws IngestionClientException, IngestionServiceException {
+        // trace ingestFromBlob
+        try (DistributedTracing.Span span = DistributedTracing.startSpan(getClientType().concat(".ingestFromBlob"), Context.NONE, ProcessKind.PROCESS,
```

Comment:
all of these are wrapped the exact same way, except for the name right? and seeing as there are more in StreamingClientIngest, might be worth refactoring to a common method/utility

Reply:
done